### PR TITLE
thrift-generator: Generate handler interface and procedures

### PR DESCRIPTION
**Pull request by bpon**
This addresses #247 by generating a handler interface and procedures for thrift services.

### CODE REVIEW

File: yarpc-thrift-generator/src/main/resources/templates/yarpc/common.stg [hunk 1]
```diff
@@ -4,26 +4,44 @@ package <context.javaPackage>;
...
 >>
 
-_method(method) ::= <<
-public io.yarpc.core.Call\<<if(method.oneway)>io.yarpc.core.transport.Ack<else><_reply(method)><endif>\> <method.javaName><_paramsWithContext(method.parameters)> {
-    <_args(method)> args = new <_args(method)><_callParams(method.parameters)>;
+_procedure(method) ::= <<
+io.yarpc.encoding.thrift.ThriftEncoding.new<if(method.oneway)>Oneway<else>Unary<endif>Procedure("<method.javaName>", new <_handler(method)>(handler), <_args(method)>.class)
+>>
+
+_clientMethod(method) ::= <<
```
Comment (abhinav) [inline yarpc-thrift-generator/src/main/resources/templates/yarpc/common.stg]: Does this format allow for comments in the template that don't get rendered? If yes, we should add some information above these sections later.

File: yarpc-thrift-generator/src/main/resources/templates/yarpc/common.stg [hunk 2]
```diff
@@ -87,18 +85,109 @@ public static final class <_reply(method)> {
...
+
+    @Override
+    public java.util.concurrent.CompletionStage\<io.yarpc.core.Reply\<<_result(method)>\>\> handleUnary(io.yarpc.core.Context ctx, <_args(method)> args) {
+        try {
+            args.validate();
+        } catch (org.apache.thrift.TException e) {
+            throw new io.yarpc.encoding.thrift.ThriftEncodingException(e);
+        }
+        return handler.<method.javaName>(ctx<if(method.parameters)>, <method.parameters: {param |args.<param.javaGetterName>()}; separator=", "><endif>)
+            .handle((reply, ex) -> {
```
Comment (abhinav) [inline yarpc-thrift-generator/src/main/resources/templates/yarpc/common.stg]: Should we handle the case where the handle function returns null?

Comment (bpon) [inline yarpc-thrift-generator/src/main/resources/templates/yarpc/common.stg]: #251

File: yarpc-thrift-generator/src/main/java/io/yarpc/thrift/generator/YarpcMethodContext.java [hunk 3]
```diff
@@ -44,12 +45,20 @@ static YarpcMethodContext fromSwiftMethodContext(
...
         .parameters(methodContext.getParameters())
         .exceptions(methodContext.getExceptions())
         .build();
   }
+
+  private static String capitalize(String str) {
+    if (str.length() > 1) {
+      return str.substring(0, 1).toUpperCase() + str.substring(1);
+    }
+    return str.toUpperCase();
```
Comment (bpon) [inline yarpc-thrift-generator/src/main/java/io/yarpc/thrift/generator/YarpcMethodContext.java]: length should be never zero, but even if it is, it'll be handled by `toUpperCase`.

Comment (abhinav) [inline yarpc-thrift-generator/src/main/java/io/yarpc/thrift/generator/YarpcMethodContext.java]: That's what I meant. It'll never be zero but you have special handling for it. If it really is zero, just returning `str` is the same as `str.toUpperCase()`.

Comment (bpon) [inline yarpc-thrift-generator/src/main/java/io/yarpc/thrift/generator/YarpcMethodContext.java]: This line is for handling length == 1.

Comment (abhinav) [inline yarpc-thrift-generator/src/main/java/io/yarpc/thrift/generator/YarpcMethodContext.java]: facepalm.jpg

My bad

File: yarpc-thrift-generator/src/main/resources/templates/yarpc/common.stg (hunk 1 above)
Comment (bpon) [inline yarpc-thrift-generator/src/main/resources/templates/yarpc/common.stg]: Comments added

**Pull request merged**